Unity 2022.3 compile errors on beta 9.6.9-beta.5 / commit 37ef016

### Summary

Installing `com.coplaydev.unity-mcp` from the moving `#beta` branch currently resolves to commit `37ef0160663d7cee0fe60e1321075e51f9a19e52` / package version `9.6.9-beta.5`.

In Unity `2022.3.4f1`, that package fails to compile before the MCP server can be used. This looks like a regression introduced by PR #1097 and then published as `9.6.9-beta.5` by PR #1099.

### Environment

- Unity: `2022.3.4f1`
- OS: Windows
- Package source:

```json
"com.coplaydev.unity-mcp": "https://github.com/CoplayDev/unity-mcp.git?path=/MCPForUnity#beta"
```

- Resolved package:
  - commit: `37ef0160663d7cee0fe60e1321075e51f9a19e52`
  - package version: `9.6.9-beta.5`

### Actual result

Unity enters Safe Mode / compile fails with these errors:

```text
Library\PackageCache\com.coplaydev.unity-mcp@37ef016066\Runtime\Helpers\UnityFindObjectsCompat.cs(66,20): error CS0103: The name 'LegacyFindObjectsOfType' does not exist in the current context

Library\PackageCache\com.coplaydev.unity-mcp@37ef016066\Runtime\Helpers\ScreenshotUtility.cs(255,28): error CS0103: The name 'CaptureFromCameraToAssetsFolder' does not exist in the current context

Library\PackageCache\com.coplaydev.unity-mcp@37ef016066\Runtime\Helpers\ScreenshotUtility.cs(260,46): error CS7036: There is no argument given that corresponds to the required formal parameter 'folderOverride' of 'ScreenshotUtility.PrepareCaptureResult(string, int, bool, string, bool)'

Library\PackageCache\com.coplaydev.unity-mcp@37ef016066\Runtime\Helpers\ScreenshotUtility.cs(273,32): error CS0103: The name 'CaptureFromCameraToAssetsFolder' does not exist in the current context

Library\PackageCache\com.coplaydev.unity-mcp@37ef016066\Runtime\Helpers\ScreenshotUtility.cs(311,45): error CS1061: 'ScreenshotCaptureResult' does not contain a definition for 'AssetsRelativePath' and no accessible extension method 'AssetsRelativePath' accepting a first argument of type 'ScreenshotCaptureResult' could be found
```

### Expected result

The Unity package should compile cleanly on Unity `2022.3.x`, or the package metadata should exclude unsupported Unity versions.

### Suspected cause

There appear to be two separate regressions.

#### 1. `UnityFindObjectsCompat.cs` preprocessor mismatch

For Unity `2022.3.x`, `FindAll(Type type, bool includeInactive)` falls into the legacy fallback branch because the new include-inactive overload is gated behind `UNITY_2023_1_OR_NEWER`.

However, the legacy reflection helpers are compiled only when `!UNITY_2022_3_OR_NEWER`, so Unity `2022.3.x` calls `LegacyFindObjectsOfType(...)` while that helper is excluded by preprocessor defines.

Unity `2022.3` documents the overload:

```csharp
Object.FindObjectsByType(Type type, FindObjectsInactive findObjectsInactive, FindObjectsSortMode sortMode)
```

So the guard probably needs to include Unity `2022.3.x`, or the legacy helper must remain compiled for the fallback branch.

Possible fix:

```diff
-#elif UNITY_2023_1_OR_NEWER
+#elif UNITY_2022_3_OR_NEWER
     return UObject.FindObjectsByType(type,
         includeInactive ? UnityEngine.FindObjectsInactive.Include : UnityEngine.FindObjectsInactive.Exclude,
         UnityEngine.FindObjectsSortMode.None);
```

#### 2. `ScreenshotUtility.cs` incomplete rename/refactor

The beta code seems to have renamed:

- `AssetsRelativePath` -> `ProjectRelativePath`
- `CaptureFromCameraToAssetsFolder` -> `CaptureFromCameraToProjectFolder`
- `CaptureToAssetsFolder` -> `CaptureToProjectFolder`
- `PrepareCaptureResult(...)` now requires `folderOverride`

But `CaptureComposited(...)` still references old names/signatures:

- `CaptureFromCameraToAssetsFolder(...)`
- `PrepareCaptureResult(fileName, superSize, ensureUniqueFileName, isAsync: false)`
- `result.AssetsRelativePath`

The same stale `Assets*` API names also appear in `Editor/Tools/ManageScene.cs`, although the Runtime assembly fails first, so Unity reports the Runtime errors before possible Editor assembly errors.

Possible fixes:

- Update remaining old call sites to the new `Project*` names and pass `folderOverride: null` where needed.
- Or keep backward-compatible wrapper members temporarily:

```csharp
public string AssetsRelativePath => ProjectRelativePath;

public static ScreenshotCaptureResult CaptureFromCameraToAssetsFolder(
    Camera camera,
    string fileName = null,
    int superSize = 1,
    bool ensureUniqueFileName = true,
    bool includeImage = false,
    int maxResolution = 0)
{
    return CaptureFromCameraToProjectFolder(
        camera, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
}
```

### Workaround

Pinning the Unity package to a fixed stable release avoids the broken beta commit:

```json
"com.coplaydev.unity-mcp": "https://github.com/CoplayDev/unity-mcp.git?path=/MCPForUnity#v9.6.8"
```

This is only a workaround; it avoids `37ef016...` / `9.6.9-beta.5` rather than fixing the beta branch.

### Related PRs / commits

- PR #1097: https://github.com/CoplayDev/unity-mcp/pull/1097
- PR #1099: https://github.com/CoplayDev/unity-mcp/pull/1099
- Failing beta commit: https://github.com/CoplayDev/unity-mcp/commit/37ef0160663d7cee0fe60e1321075e51f9a19e52

--- a/.github/workflows/beta-release.yml
+++ b/.github/workflows/beta-release.yml
@@ -13,9 +13,21 @@ on:
       - "MCPForUnity/**"
 
 jobs:
+  unity_tests:
+    name: Unity tests gate
+    if: github.actor != 'github-actions[bot]'
+    uses: ./.github/workflows/unity-tests.yml
+    secrets: inherit
+
+  python_tests:
+    name: Python tests gate
+    if: github.actor != 'github-actions[bot]'
+    uses: ./.github/workflows/python-tests.yml
+
   update_unity_beta_version:
     name: Update Unity package to beta version
     runs-on: ubuntu-latest
+    needs: [unity_tests, python_tests]
     # Avoid running when the workflow's own automation merges the PR
     # created by this workflow (prevents a version-bump loop).
     if: github.actor != 'github-actions[bot]'
@@ -143,6 +155,7 @@ jobs:
   publish_pypi_prerelease:
     name: Publish beta to PyPI (pre-release)
     runs-on: ubuntu-latest
+    needs: [unity_tests, python_tests]
     # Avoid double-publish when the bot merges the version bump PR
     if: github.actor != 'github-actions[bot]'
     environment:

--- a/.github/workflows/python-tests.yml
+++ b/.github/workflows/python-tests.yml
@@ -6,7 +6,19 @@ on:
     paths:
       - Server/**
       - .github/workflows/python-tests.yml
+  pull_request:
+    branches: [main, beta]
+    paths:
+      - Server/**
+      - .github/workflows/python-tests.yml
   workflow_dispatch: {}
+  workflow_call:
+    inputs:
+      ref:
+        description: "Git ref to test (defaults to the triggering ref)."
+        type: string
+        required: false
+        default: ""
 
 jobs:
   test:
@@ -15,6 +27,8 @@ jobs:
     steps:
       - name: Checkout repository
         uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.ref || github.ref }}
 
       - name: Install uv
         uses: astral-sh/setup-uv@v4

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -19,9 +19,23 @@ on:
         required: true
 
 jobs:
+  unity_tests:
+    name: Unity tests gate
+    uses: ./.github/workflows/unity-tests.yml
+    with:
+      ref: beta
+    secrets: inherit
+
+  python_tests:
+    name: Python tests gate
+    uses: ./.github/workflows/python-tests.yml
+    with:
+      ref: beta
+
   bump:
     name: Bump version, tag, and create release
     runs-on: ubuntu-latest
+    needs: [unity_tests, python_tests]
     permissions:
       contents: write
       pull-requests: write

--- a/.github/workflows/unity-tests.yml
+++ b/.github/workflows/unity-tests.yml
@@ -31,6 +31,8 @@ jobs:
   testAllModes:
     name: Test in ${{ matrix.testMode }}
     runs-on: ubuntu-latest
+    permissions:
+      contents: read
     if: >
       github.event_name != 'pull_request_target' ||
       (github.event.pull_request.head.repo.full_name != github.repository &&
@@ -50,6 +52,7 @@ jobs:
         with:
           lfs: true
           ref: ${{ inputs.ref || github.event.pull_request.head.sha || github.ref }}
+          persist-credentials: false
 
       - name: Detect Unity license secrets
         id: detect
@@ -130,7 +133,7 @@ jobs:
           fi
 
       - uses: actions/upload-artifact@v4
-        if: always() && steps.detect.outputs.unity_ok == 'true'
+        if: always() && steps.detect.outputs.unity_ok == 'true' && steps.tests.outcome != 'skipped'
         with:
           name: Test results for ${{ matrix.testMode }}
           path: ${{ steps.tests.outputs.artifactsPath }}

--- a/.github/workflows/unity-tests.yml
+++ b/.github/workflows/unity-tests.yml
@@ -2,8 +2,26 @@ name: Unity Tests
 
 on:
   workflow_dispatch: {}
+  workflow_call:
+    inputs:
+      ref:
+        description: "Git ref to test (defaults to the triggering ref)."
+        type: string
+        required: false
+        default: ""
   push:
     branches: ["**"]
+    paths:
+      - TestProjects/UnityMCPTests/**
+      - MCPForUnity/Editor/**
+      - .github/workflows/unity-tests.yml
+  # Fork PRs: maintainer applies the 'safe-to-test' label after reviewing
+  # the diff. The workflow runs with UNITY_LICENSE in scope against the
+  # PR's head SHA. Re-pushed commits do NOT auto-trigger — maintainer must
+  # remove and re-apply the label to re-run after additional review.
+  pull_request_target:
+    types: [labeled]
+    branches: [main, beta]
     paths:
       - TestProjects/UnityMCPTests/**
       - MCPForUnity/Editor/**
@@ -13,6 +31,10 @@ jobs:
   testAllModes:
     name: Test in ${{ matrix.testMode }}
     runs-on: ubuntu-latest
+    if: >
+      github.event_name != 'pull_request_target' ||
+      (github.event.pull_request.head.repo.full_name != github.repository &&
+       github.event.label.name == 'safe-to-test')
     strategy:
       fail-fast: false
       matrix:
@@ -27,6 +49,7 @@ jobs:
         uses: actions/checkout@v4
         with:
           lfs: true
+          ref: ${{ inputs.ref || github.event.pull_request.head.sha || github.ref }}
 
       - name: Detect Unity license secrets
         id: detect

--- a/MCPForUnity/Editor/Tools/ManageScene.cs
+++ b/MCPForUnity/Editor/Tools/ManageScene.cs
@@ -598,11 +598,14 @@ namespace MCPForUnity.Editor.Tools
                 {
                     if (!Application.isBatchMode) EnsureGameView();
 
+                    string folderOverride = ScreenshotPreferences.Resolve(cmd.outputFolder);
                     ScreenshotCaptureResult result = ScreenshotUtility.CaptureFromCameraToProjectFolder(
                         targetCamera, fileName, resolvedSuperSize, ensureUniqueFileName: true,
-                        includeImage: includeImage, maxResolution: maxResolution);
+                        includeImage: includeImage, maxResolution: maxResolution,
+                        folderOverride: folderOverride);
 
-                    AssetDatabase.ImportAsset(result.ProjectRelativePath, ImportAssetOptions.ForceSynchronousImport);
+                    if (ScreenshotUtility.IsUnderAssets(result.ProjectRelativePath))
+                        AssetDatabase.ImportAsset(result.ProjectRelativePath, ImportAssetOptions.ForceSynchronousImport);
                     string message = $"Screenshot captured to '{result.ProjectRelativePath}' (camera: {targetCamera.name}).";
                     return new SuccessResponse(message, BuildScreenshotResponseData(result, targetCamera.name, includeImage));
                 }
@@ -611,11 +614,14 @@ namespace MCPForUnity.Editor.Tools
                 {
                     if (!Application.isBatchMode) EnsureGameView();
 
+                    string folderOverride = ScreenshotPreferences.Resolve(cmd.outputFolder);
                     ScreenshotCaptureResult result = ScreenshotUtility.CaptureComposited(
                         fileName, resolvedSuperSize, ensureUniqueFileName: true,
-                        includeImage: true, maxResolution: maxResolution);
+                        includeImage: true, maxResolution: maxResolution,
+                        folderOverride: folderOverride);
 
-                    AssetDatabase.ImportAsset(result.ProjectRelativePath, ImportAssetOptions.ForceSynchronousImport);
+                    if (ScreenshotUtility.IsUnderAssets(result.ProjectRelativePath))
+                        AssetDatabase.ImportAsset(result.ProjectRelativePath, ImportAssetOptions.ForceSynchronousImport);
                     string cameraName = Camera.main != null ? Camera.main.name : "composited";
                     string message = $"Screenshot captured to '{result.ProjectRelativePath}' (camera: {cameraName}).";
                     return new SuccessResponse(message, BuildScreenshotResponseData(result, cameraName, includeImage: true));

--- a/MCPForUnity/Editor/Tools/ManageScene.cs
+++ b/MCPForUnity/Editor/Tools/ManageScene.cs
@@ -598,12 +598,12 @@ namespace MCPForUnity.Editor.Tools
                 {
                     if (!Application.isBatchMode) EnsureGameView();
 
-                    ScreenshotCaptureResult result = ScreenshotUtility.CaptureFromCameraToAssetsFolder(
+                    ScreenshotCaptureResult result = ScreenshotUtility.CaptureFromCameraToProjectFolder(
                         targetCamera, fileName, resolvedSuperSize, ensureUniqueFileName: true,
                         includeImage: includeImage, maxResolution: maxResolution);
 
-                    AssetDatabase.ImportAsset(result.AssetsRelativePath, ImportAssetOptions.ForceSynchronousImport);
-                    string message = $"Screenshot captured to '{result.AssetsRelativePath}' (camera: {targetCamera.name}).";
+                    AssetDatabase.ImportAsset(result.ProjectRelativePath, ImportAssetOptions.ForceSynchronousImport);
+                    string message = $"Screenshot captured to '{result.ProjectRelativePath}' (camera: {targetCamera.name}).";
                     return new SuccessResponse(message, BuildScreenshotResponseData(result, targetCamera.name, includeImage));
                 }
 
@@ -615,9 +615,9 @@ namespace MCPForUnity.Editor.Tools
                         fileName, resolvedSuperSize, ensureUniqueFileName: true,
                         includeImage: true, maxResolution: maxResolution);
 
-                    AssetDatabase.ImportAsset(result.AssetsRelativePath, ImportAssetOptions.ForceSynchronousImport);
+                    AssetDatabase.ImportAsset(result.ProjectRelativePath, ImportAssetOptions.ForceSynchronousImport);
                     string cameraName = Camera.main != null ? Camera.main.name : "composited";
-                    string message = $"Screenshot captured to '{result.AssetsRelativePath}' (camera: {cameraName}).";
+                    string message = $"Screenshot captured to '{result.ProjectRelativePath}' (camera: {cameraName}).";
                     return new SuccessResponse(message, BuildScreenshotResponseData(result, cameraName, includeImage: true));
                 }
 
@@ -747,7 +747,7 @@ namespace MCPForUnity.Editor.Tools
         {
             var data = new Dictionary<string, object>
             {
-                { "path", result.AssetsRelativePath },
+                { "path", result.ProjectRelativePath },
                 { "fullPath", result.FullPath },
                 { "superSize", result.SuperSize },
                 { "isAsync", false },

--- a/MCPForUnity/Runtime/Helpers/ScreenshotUtility.cs
+++ b/MCPForUnity/Runtime/Helpers/ScreenshotUtility.cs
@@ -252,12 +252,12 @@ namespace MCPForUnity.Runtime.Helpers
             {
                 var fallbackCamera = FindAvailableCamera();
                 if (fallbackCamera != null)
-                    return CaptureFromCameraToAssetsFolder(fallbackCamera, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
+                    return CaptureFromCameraToProjectFolder(fallbackCamera, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
 
                 throw new InvalidOperationException("ScreenCapture module is unavailable and no fallback camera found.");
             }
 
-            ScreenshotCaptureResult result = PrepareCaptureResult(fileName, superSize, ensureUniqueFileName, isAsync: false);
+            ScreenshotCaptureResult result = PrepareCaptureResult(fileName, superSize, ensureUniqueFileName, folderOverride: null, isAsync: false);
             Texture2D tex = null;
             Texture2D downscaled = null;
             string imageBase64 = null;
@@ -270,7 +270,7 @@ namespace MCPForUnity.Runtime.Helpers
                     // Fallback to camera-based if ScreenCapture fails
                     var cam = FindAvailableCamera();
                     if (cam != null)
-                        return CaptureFromCameraToAssetsFolder(cam, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
+                        return CaptureFromCameraToProjectFolder(cam, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
                     throw new InvalidOperationException("ScreenCapture.CaptureScreenshotAsTexture returned null and no fallback camera available.");
                 }
 
@@ -308,7 +308,7 @@ namespace MCPForUnity.Runtime.Helpers
             if (includeImage && imageBase64 != null)
             {
                 return new ScreenshotCaptureResult(
-                    result.FullPath, result.AssetsRelativePath, result.SuperSize, false,
+                    result.FullPath, result.ProjectRelativePath, result.SuperSize, false,
                     imageBase64, imgW, imgH);
             }
             return result;

--- a/MCPForUnity/Runtime/Helpers/ScreenshotUtility.cs
+++ b/MCPForUnity/Runtime/Helpers/ScreenshotUtility.cs
@@ -246,18 +246,20 @@ namespace MCPForUnity.Runtime.Helpers
             int superSize = 1,
             bool ensureUniqueFileName = true,
             bool includeImage = false,
-            int maxResolution = 0)
+            int maxResolution = 0,
+            string folderOverride = null)
         {
             if (!IsScreenCaptureModuleAvailable)
             {
                 var fallbackCamera = FindAvailableCamera();
                 if (fallbackCamera != null)
-                    return CaptureFromCameraToProjectFolder(fallbackCamera, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
+                    return CaptureFromCameraToProjectFolder(fallbackCamera, fileName, superSize, ensureUniqueFileName,
+                        includeImage, maxResolution, folderOverride: folderOverride);
 
                 throw new InvalidOperationException("ScreenCapture module is unavailable and no fallback camera found.");
             }
 
-            ScreenshotCaptureResult result = PrepareCaptureResult(fileName, superSize, ensureUniqueFileName, folderOverride: null, isAsync: false);
+            ScreenshotCaptureResult result = PrepareCaptureResult(fileName, superSize, ensureUniqueFileName, folderOverride: folderOverride, isAsync: false);
             Texture2D tex = null;
             Texture2D downscaled = null;
             string imageBase64 = null;
@@ -270,7 +272,8 @@ namespace MCPForUnity.Runtime.Helpers
                     // Fallback to camera-based if ScreenCapture fails
                     var cam = FindAvailableCamera();
                     if (cam != null)
-                        return CaptureFromCameraToProjectFolder(cam, fileName, superSize, ensureUniqueFileName, includeImage, maxResolution);
+                        return CaptureFromCameraToProjectFolder(cam, fileName, superSize, ensureUniqueFileName,
+                            includeImage, maxResolution, folderOverride: folderOverride);
                     throw new InvalidOperationException("ScreenCapture.CaptureScreenshotAsTexture returned null and no fallback camera available.");
                 }
 

